src/filesystem/impls/filer/UrlCache.js

```diff
@@ -0,0 +1,381 @@
+        var prefix = brackets.env === "production" ? "/dist" : "/src";
+        this.generateVFSUrlForPath = function(path) {
+            var a = document.createElement("a");
+            a.href = prefix + "/vfs" + path;
+            return a.href;
+        };
+
+        this.projectCacheName = Path.join("vfs", StartupState.project("root"));
+        this.baseUrl = this.generateVFSUrlForPath(StartupState.project("root")) + "/";
+        this.shouldRewriteUrls = false;
+    }
+    CacheStorageUrlProvider.prototype = Object.create(UrlProviderBase.prototype);
+    CacheStorageUrlProvider.prototype.constructor = CacheStorageUrlProvider;
+
+    CacheStorageUrlProvider.prototype.clear = function(callback) {
+        var self = this;
+        var projectCacheName = self.projectCacheName;
+
+        // Delete existing cache for this root, and recreate empty cache.
+        window.caches
+            .delete(projectCacheName)
+            .then(function() {
+                caches.open(projectCacheName).then(function() {
+                    self.paths = {};
+                    self.urls = {};
+                    callback();
+                });
+            })
+            .catch(callback);
+    };
+    CacheStorageUrlProvider.prototype.init = function(callback) {
+        this.clear(callback);
+    };
+    CacheStorageUrlProvider.prototype.createURL = function(filename, blob, type, callback) {
+        var response = new Response(blob, {
+            status: 200,
+            statusText: "Served from Thimble's Offline Cache"
+        });
+
+        var headers = new Headers();
+        headers.append("Content-Type", type);
+
+        var url = this.generateVFSUrlForPath(filename);
+        var request = new Request(url, {
+            method: "GET",
+            headers: headers
+        });
+
+        this.urls[filename] = url;
+        this.paths[url] = filename;
+
+        window.caches
+            .open(this.projectCacheName)
+            .then(function(cache) {
+                cache.put(request, response).then(function() {
+                    callback(null, url);
+                });
+            })
+            .catch(callback);
+    };
+    CacheStorageUrlProvider.prototype.remove = function(path, callback) {
+        var self = this;
+        var removed = [];
+
+        function _maybeRemove(pathPart) {
+            var deferred = new $.Deferred();
+            var url = self.urls[pathPart];
+
+            // The first time a file is written, we won't have
+            // a stale cache entry to clean up.
+            if(!url) {
+                return deferred.resolve().promise();
+            }
+
+            // If this filename matches exactly, or is a root path (i.e., other
+            // filenames begin with "<path>/...", remove it. Otherwise just skip.
+            if(pathPart === path || pathPart.indexOf(path + "/") === 0) {
+                removed.push(pathPart);
+
+                delete self.urls[pathPart];
```

Comment:
That is likely a bug in both of them.  I'll file it as a follow up to fix later vs. blocking this on it.